central/resourcecollection/service/service_impl.go

```diff
@@ -83,3 +94,58 @@ func (s *serviceImpl) getCollection(ctx context.Context, id string) (*v1.GetColl
 		Deployments: nil,
 	}, nil
 }
+
+// CreateCollection creates a new collection from the given request
+func (s *serviceImpl) CreateCollection(ctx context.Context, request *v1.CreateCollectionRequest) (*v1.CreateCollectionResponse, error) {
+	if !features.ObjectCollections.Enabled() {
+		return nil, nil
+	}
+
+	if request.GetName() == "" {
+		return nil, errors.Wrap(errox.InvalidArgs, "Collection name should not be empty")
+	}
+
+	collection, err := collectionRequestToCollection(ctx, request, true)
+	if err != nil {
+		return nil, err
+	}
+
+	err = s.datastore.AddCollection(ctx, collection)
+	if err != nil {
+		return nil, err
+	}
+
+	return &v1.CreateCollectionResponse{Collection: collection}, nil
+}
+
+func collectionRequestToCollection(ctx context.Context, request collectionRequest, isCreate bool) (*storage.ResourceCollection, error) {
+	user := utils.UserFromContext(ctx)
+	if user == nil {
+		return nil, errors.New("User identity not provided")
+	}
+
+	timeNow := protoconv.ConvertTimeToTimestamp(time.Now())
+
+	collection := &storage.ResourceCollection{
+		Name:              request.GetName(),
+		Description:       request.GetDescription(),
+		LastUpdated:       timeNow,
+		UpdatedBy:         user,
+		ResourceSelectors: request.GetResourceSelectors(),
+	}
```

Comment:
You might have to do some digging to figure out a clean way of doing this (maybe [generics](https://go.dev/doc/tutorial/generics)!) but for future work we need to also include the `Id` field population in the object creation for Update requests

Reply:
Ah yes, I will add an `idGetter func() string` to the arguments, and update can then pass `func() {return  request.GetId() } ` as that func.

About generics, I thought to do it using generics, but since we are calling functions `GetName`, `GetDescription`, etc on the passed obj, if I call them on generic type R, compilation would fail. The solution for that is to constrain the generic type using an interface. In this case the constraining interface would look like

```
type collectionRequest interface {
    *v1.CreateCollectionRequest | *v1.DryRunCollectionRequest | *v1.UpdateCollectionRequest
}

func [R collectionRequest] collectionRequestToCollection(ctx context, request R, isCreate bool) {...}
```

However, this fails to compile too since go 1.18 doesn't have a good support for union of structs within a constraining interface. Since  *v1.CreateCollectionRequest and others are structs and not interfaces themselves, `GetName` are yet not accessible from collectionRequest.

The workaround for this is to then add those methods to the constraining interface as well, which makes the interface look like
```
type collectionRequest interface {
	*v1.CreateCollectionRequest

	GetName() string
	GetDescription() string
	GetResourceSelectors() []*storage.ResourceSelector
	GetEmbeddedCollectionIds() []string
}
```

This works, but then I still had to put those methods in the interface. so I just avoided the generics altogether for this case.